ssh3: init at 0.1.7, nixos/ssh3: init

RaitoBezarius opened the pull request:
## Description of changes

https://github.com/francoismichel/ssh3

This is an experimental software, please do not go and shill it to others if you don't understand the protocol behind.
This didn't receive any formal review, there's a lot of trust involved.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RaitoBezarius committed:
ssh3: init at 0.1.7

https://github.com/francoismichel/ssh3

RaitoBezarius committed:
nixos/ssh3-server: init

Co-authored-by: Tom Hubrecht <Tom-Hubrecht@users.noreply.github.com>

RaitoBezarius committed:
nixos/tests/ssh3: init

SuperSandro2000 commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,77 @@
+        description = lib.mdDoc "Domain name for this server";
+      };
+
+      # Secret path
```
```suggestion
```
Only creates confusion with the description

SuperSandro2000 reviewed (commented):
Rest looks good to me

mweinelt commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,77 @@
+            };
+          };
+        });
+        default = [ { addr = "0.0.0.0"; port = 22; } { addr = "[::1]"; port = 22; } ];
```
```suggestion
        default = [ { addr = "0.0.0.0"; port = 22; } { addr = "[::]"; port = 22; } ];
```

RaitoBezarius replied:
thank you ser

RaitoBezarius committed:
nixos/bind: fix missing defaults

The builder function for zone marks them as default but the module system
has no way to know this.

So we will make them default also in the module system.

This fixes an issue when you use the ACME test server setup without defining yourself
all your zones manually.

RaitoBezarius committed:
nixos/tests/common/resolver: fix BIND DNS zones

Previously, zones supposedly were list, now they are attributes.
We do the missed (?) migration.

RaitoBezarius committed:
nixos/tests/common/resolver: Aliases should be accumulated in a flat list

Previously, they were accumulated in a nested list, which is arguably a bug
and a typo between `optional` and `optionals`.

drupol commented on pkgs/by-name/ss/ssh3/package.nix:
```diff
@@ -0,0 +1,33 @@
+  ldflags = [ "-s" "-w" ];
+
+  meta = with lib; {
+    description = "Revisit of the SSH protocol using QUIC + TLS 1.3 and HTTP authorization";
```
```suggestion
    description = "SSH3: faster and rich secure shell using HTTP/3";
```

RaitoBezarius replied:
I don't recommend adopting upstream "faster and rich secure shell" representation. This is a preprint paper headline, not a faithful description of what it is right now.

drupol replied:
Fair enough!

viraptor commented on pkgs/by-name/ss/ssh3/package.nix:
```diff
@@ -0,0 +1,38 @@
+
+  postInstall = ''
+    mv $out/bin/client $out/bin/ssh3
+    mv $out/bin/server $out/bin/ssh3-server
```
On darwin, only the client is supported so far and the build currently fails with:
```
Building subPackage ./cli/server
# ssh3/util/linux_util
util/linux_util/linux_user.go:32:10: fatal error: 'shadow.h' file not found
#include <shadow.h>
```
Could you disable the server subpackage on `isDarwin`?

RaitoBezarius replied:
If you can provide me with a patch to do so, yes.

drupol replied:
I think this issue is linked: https://github.com/francoismichel/ssh3/pull/56

viraptor replied:
Nice, yes, that's probably a better solution than extracting the client build today.

RaitoBezarius committed:
rl-2405: add `ssh3` NixOS module

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

Tom-Hubrecht commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,94 @@
+{ pkgs, config, lib, ... }:
+let
+  inherit (lib) mkEnableOption mkOption mkPackageOptionMD mkIf types;
```
My ocd would like to see those sorted....

RaitoBezarius replied:
Your OCD is really bad, I need to put them on each line to make this realistically sortable using `vim`, otherwise, it's hellish.

Tom-Hubrecht commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,94 @@
+        # Keep 22 or 443... ?
+        default = [ { addr = "[::]"; port = 443; } ];
+        example = [ { addr = "192.168.3.1"; port = 22; } { addr = "0.0.0.0"; port = 64022; } ];
+        description = lib.mdDoc ''
```
Isn't `lib.mdDoc` the identity function ?

RaitoBezarius replied:
Done.

Tom-Hubrecht commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,94 @@
+        after = [ "network.target" "acme-finished-${cfg.domain}.target" ];
+
+        serviceConfig = {
+          ExecStart = "${cfg.package}/bin/ssh3-server ${listenParameters} ${certificateParameters} -url-path ${cfg.urlPath}";
```
```suggestion
          ExecStart = "${lib.getExe' cfg.package "ssh3-server"} ${listenParameters} ${certificateParameters} -url-path ${cfg.urlPath}";
```

RaitoBezarius replied:
Done.

Tom-Hubrecht commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,94 @@
+              '';
+            };
+            port = mkOption {
+              type = types.int;
```
```suggestion
              type = types.port;
```

RaitoBezarius replied:
Done.

Tom-Hubrecht commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,94 @@
+let
+  inherit (lib) mkEnableOption mkOption mkPackageOptionMD mkIf types;
+  cfg = config.services.ssh3-server;
+  acmeDirectory = config.security.acme.certs."${cfg.domain}".directory;
```
```suggestion
  acmeDirectory = config.security.acme.certs.${cfg.domain}.directory;
```

RaitoBezarius replied:
Done.

Tom-Hubrecht commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,94 @@
+        after = [ "network.target" "acme-finished-${cfg.domain}.target" ];
+
+        serviceConfig = {
+          ExecStart = "${cfg.package}/bin/ssh3-server ${listenParameters} ${certificateParameters} -url-path ${cfg.urlPath}";
```
You can use `escapeSystemdExecArgs` and avoid string manipulations

RaitoBezarius replied:
Wow, this needs to think. I will not do this right now.

RaitoBezarius commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,94 @@
+        type = with types; listOf (submodule {
+          options = {
+            addr = mkOption {
+              type = types.str;
```
I used the scope.

RaitoBezarius reviewed (commented):
@Tom-Hubrecht PTAL modulo systemd escape.

Tom-Hubrecht commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,99 @@
+
+    config = mkIf cfg.enable {
+      systemd.services.ssh3-server =
+      let
+        listenParameters = lib.concatStringsSep " " (map ({ addr, port }: "-bind ${addr}:${toString port}") cfg.listenAddresses);
+        certificateParameters = "-cert ${cfg.certificate.publicKeyPath} -key ${cfg.certificate.privateKeyPath}";
+      in
+      {
+        description = "SSH3 Daemon";
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" "acme-finished-${cfg.domain}.target" ];
+
+        serviceConfig = {
+          ExecStart = "${lib.getExe' cfg.package "ssh3-server"} ${listenParameters} ${certificateParameters} -url-path ${cfg.urlPath}";
```
For using `utils.escapeSystemdExecArgs`, you can do : 

```suggestion
      {
        description = "SSH3 Daemon";
        wantedBy = [ "multi-user.target" ];
        after = [ "network.target" "acme-finished-${cfg.domain}.target" ];

        serviceConfig = {
          ExecStart = escapeSystemdExecArgs ([
            (lib.getExe' cfg.package "ssh3-server")
             "-cert" cfg.certificate.publicKeyPath
             "-key" cfg.certificate.privateKeyPath
             "-url-path" cfg.urlPath
           ] ++ (concatLists (map ({ addr, port }: [ "-bind" "${addr}:${port}" ]) cfg.listenAddresses));
```

Tom-Hubrecht commented on nixos/modules/services/networking/ssh/ssh3-server.nix:
```diff
@@ -0,0 +1,99 @@
+
+      package = mkPackageOptionMD pkgs "ssh3" { };
+      listenAddresses = mkOption {
+        type = with types; listOf (submodule {
+          options = {
+            addr = mkOption {
+              type = str;
+              description = ''
+                Host, IPv4 or IPv6 address to listen to.
+              '';
+            };
+            port = mkOption {
+              type = port;
+              description = ''
+                Port to listen to.
+              '';
+            };
+          };
```
Actually, I don't see `addr` or `port` used individually, it might be better to just make it a list of strings of the form `addr:port`

RaitoBezarius replied:
This means that you need to reparse the data if you need to introspect config for some reason. I would find this a bit bizarre, is it a nitpick or a blocker in your eyes? :P

Tom-Hubrecht replied:
Not a blocker, just a note

RaitoBezarius replied:
Resolving for now, I don't think it's very important to solve

Tom-Hubrecht commented on pkgs/by-name/ss/ssh3/package.nix:
```diff
@@ -0,0 +1,33 @@
+  vendorHash = "sha256-VUNvb7m1nnH+mXUsnIKyPKJEVSMXBAaS4ihi5DZeFiI=";
+
+  ldflags = [ "-s" "-w" ];
+
```
Maybe add the nixos test as a `passthru.tests` ?

RaitoBezarius committed:
ssh3: use `nixosTests.ssh3` as a basic test

This helps to ensure that `ssh3` works across upgrades.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius commented:
The NixOS test is failing at the moment.

Aleksanaa commented:
Hi Raito, would you like to give it another attempt?

RaitoBezarius commented:
> Hi Raito, would you like to give it another attempt?

No, due to the current governance situation, it's out of the question. I am still watching how the situation is unfolding, but it's clearly very disappointing and an enormous shitshow.

Aleksanaa commented:
Then could I take this diff and submit again?

SuperSandro2000 commented:
The repo and it's contributions are licensed under MIT, so as long as you attribute you are allowed to do that. Git's Co-Author-By field is a very good way to attribute multiple people to commits.

RaitoBezarius commented:
> Then could I take this diff and submit again?

I would rather not, this sort of software requires expertise and I am not exactly sure it's sufficient just to rerun the PR is a good solution to have a good quality packaging of this.

Feel free to cherry pick the random fixes for the BIND VM tests stuff, though, they are useful in general.

If I am to be `Co-authored-by` against my will of seeing this being upstreamed in this state, please add a line about how I did not want this in the commit message.
